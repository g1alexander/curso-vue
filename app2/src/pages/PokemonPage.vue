<script setup lang="ts">
import PokemonPicture from "../components/PokemonPicture.vue";
import PokemonOptions from "../components/PokemonOptions.vue";
import { getPokemonsOptions } from "../helpers/getPokemonOptions";
import { onMounted, provide, ref, watchEffect } from "vue";
import type { PokemonsOptions } from "@/helpers/interfaces/PokemonOptions";

const pokemonRes = ref<PokemonsOptions[]>([]);
const pokemon = ref<PokemonsOptions>({ name: "", id: 1 });
const pokemonId = ref<number>(0);
const message = ref<string>("");
const showPokemon = ref<boolean>(false);

provide("pokemon", { pokemonId, showPokemon });

watchEffect(() => {
  if (pokemonId.value === pokemon.value.id) {
    message.value = `You already have this pokemon ${pokemon.value.name}`;
  } else {
    message.value = `F parce era ${pokemon.value.name}`;
  }
});

const mountPokemon = async () => {
  pokemonRes.value = await getPokemonsOptions();

  const pokemonId: number = Math.floor(Math.random() * pokemonRes.value.length);

  pokemon.value = pokemonRes.value[pokemonId];
};

onMounted(() => {
  mountPokemon();
});

const newGame = async () => {
  message.value = "";
  pokemonId.value = 0;
  showPokemon.value = false;
  mountPokemon();
};
</script>

<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>

  <section v-else>
    <h1>Â¿Quien es el pokemon? {{ pokemonId }}</h1>

    <PokemonPicture :pokemon-id="pokemon.id" :show-pokemon="showPokemon" />

    <PokemonOptions :pokemons="pokemonRes" />

    <template v-if="showPokemon">
      <h2>{{ message }}</h2>
      <button @click="newGame">Nuevo juego</button>
    </template>
  </section>
</template>

<style></style>
