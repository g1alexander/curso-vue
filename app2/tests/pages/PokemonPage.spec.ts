import PokemonPage from "@/pages/PokemonPage.vue";
import { shallowMount, VueWrapper } from "@vue/test-utils";
import { beforeEach, describe, expect, test, spyOn } from "vitest";

describe("component PokemonPage", () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);
  });

  test("debe de hacer match con el snapshot", () => {
    // esta prueba es ineccesario debido a que el snapshot esta vacio
    // porque no hay pokemon
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("debe de llamar al mountPokemon", () => {
    //la prueba pasa pero esta mal hecha
    const wrapper = shallowMount(PokemonPage);

    const obj = {
      mountPokemon: wrapper.vm.mountPokemon,
    };

    const spy = spyOn(obj, "mountPokemon");

    obj.mountPokemon();

    expect(spy).toHaveBeenCalled();
  });
});
