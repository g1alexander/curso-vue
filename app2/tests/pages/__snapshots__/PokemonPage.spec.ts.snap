// Vitest Snapshot v1

exports[`component PokemonPage > debe de hacer match con el snapshot 1`] = `"<h1>Espere por favor...</h1>"`;
